File: shared/Harbor/v2/harbor_py/harbor_py/server/env.py
"""Environment-aware HTTP client factory for server-side use.

Selects the underlying transport based on the ``NODE_ENV`` environment variable
so that callers never need to branch on the environment themselves:

- **development** (default when ``NODE_ENV`` is absent or not ``"production"``)
  — uses a plain :class:`~harbor_py.client.HarborClient` backed by ``httpx``.
- **production** (``NODE_ENV == "production"``) — uses an
  :class:`~harbor_py.server.iam.IamHarborClient` that automatically attaches a
  Google Cloud OIDC identity token to every outgoing request, enabling secure
  service-to-service calls on Google Cloud Run.

**Server-only** — this module should never be used in browser code.
"""

from __future__ import annotations

import os

from harbor_py.client import HarborClient, create_harbor_client
from harbor_py.server.iam import IamHarborClient, create_iam_harbor_client

# ─── Type alias ───────────────────────────────────────────────────────────────

#: Return type of :func:`create_server_harbor_client`.  In development this is
#: a plain :class:`~harbor_py.client.HarborClient`; in production an
#: :class:`~harbor_py.server.iam.IamHarborClient`.  Both expose the same
#: async HTTP interface so callers can use either transparently.
ServerHarborClient = HarborClient | IamHarborClient


# ─── Factory ──────────────────────────────────────────────────────────────────


def create_server_harbor_client(
    base_url: str,
    audience: str | None = None,
    timeout_seconds: float | None = 30.0,
    default_headers: dict[str, str] | None = None,
) -> ServerHarborClient:
    """Create an environment-aware HTTP client for server-side service calls.

    The underlying transport is chosen automatically based on ``NODE_ENV``:

    - In **development** (default): plain ``httpx`` client — no authentication
      overhead.
    - In **production**: IAM-authenticated ``httpx`` client — attaches a
      Google Cloud OIDC identity token to every request.

    Callers use the returned client identically in both environments::

        client = create_server_harbor_client(base_url=config.users_api_url)

        async with client as c:
            response = await c.get("/api/v1/users")

    Args:
        base_url: Base URL prepended to every request path.
        audience: IAM OIDC audience used in production mode.  For Cloud Run
            services this is typically the service's HTTPS URL, e.g.
            ``https://my-service-hash-uc.a.run.app``.  When omitted,
            ``base_url`` is used as the audience.  Has no effect in
            development mode.
        timeout_seconds: Per-request timeout in seconds.  Pass ``None`` to
            disable.  Defaults to ``30.0``.
        default_headers: Headers attached to every outgoing request.

    Returns:
        A :class:`ServerHarborClient` (either :class:`~harbor_py.client.HarborClient`
        or :class:`~harbor_py.server.iam.IamHarborClient`) ready for use.
    """
    if os.environ.get("NODE_ENV") == "production":
        return create_iam_harbor_client(
            base_url=base_url,
            audience=audience or base_url,
            timeout_seconds=timeout_seconds,
            default_headers=default_headers,
        )

    return create_harbor_client(
        base_url=base_url,
        timeout_seconds=timeout_seconds,
        default_headers=default_headers,
    )

File: shared/Harbor/v2/harbor_ts/src/server/env.ts
/**
 * Environment-aware HTTP client factory for server-side use.
 *
 * Selects the underlying transport based on the `NODE_ENV` environment
 * variable so that callers never need to branch on the environment themselves:
 *
 * - **development** (default when `NODE_ENV` is absent or not `"production"`):
 *   Uses a plain axios client ({@link createHarborClient}).
 * - **production** (`NODE_ENV === "production"`): Uses an IAM-authenticated
 *   axios client ({@link createIamHarborClient}) that automatically attaches a
 *   Google Cloud OIDC identity token to every outgoing request.
 *
 * This file is **server-only** — never import it in browser/frontend code.
 */

import { createHarborClient } from '../client/index.js';
import type { HarborClient, HarborClientOptions } from '../client/index.js';
import { createIamHarborClient } from './iam.js';

// ─── Types ────────────────────────────────────────────────────────────────────

/**
 * Configuration options for {@link createServerHarborClient}.
 *
 * Extends {@link HarborClientOptions} with an optional `audience` field that
 * is only used in production (for Google Cloud IAM authentication).  When
 * `audience` is omitted in production the `baseURL` is used as the audience.
 */
export interface ServerHarborClientOptions extends HarborClientOptions {
  /**
   * IAM OIDC audience used in production mode.
   *
   * For Cloud Run services this is typically the service's HTTPS URL, e.g.
   * `https://my-service-hash-uc.a.run.app`.
   *
   * When omitted, the `baseURL` is used as the audience.  This field has no
   * effect in development mode.
   */
  audience?: string;
}

// ─── Factory ──────────────────────────────────────────────────────────────────

/**
 * Creates a pre-configured {@link HarborClient} whose transport is chosen
 * automatically based on `NODE_ENV`:
 *
 * - In **development** (default): plain axios — no authentication overhead.
 * - In **production**: IAM-authenticated axios — attaches a Google Cloud OIDC
 *   identity token to every request, enabling secure service-to-service calls
 *   on Google Cloud Run.
 *
 * Callers use the returned client identically in both environments:
 * ```typescript
 * const client = createServerHarborClient({
 *   baseURL: config.usersApiServiceUrl,
 *   timeoutMs: config.requestTimeoutMs,
 * });
 *
 * const response = await client.get('/api/v1/users');
 * ```
 *
 * @param options - Client configuration.  Set `audience` to override the IAM
 *   OIDC audience used in production (defaults to `baseURL`).
 * @returns A configured HarborClient ready for use.
 *
 * @deprecated The procedural factory interface (`createServerHarborClient`) is deprecated.
 * Use the object-oriented `ServerHarborClient` class from
 * `@elastic-resume-base/harbor` v3 instead:
 * ```typescript
 * import { ServerHarborClient } from '@elastic-resume-base/harbor/server'; // v3
 * const client = new ServerHarborClient({ baseURL: '...', timeoutMs: 30_000 });
 * ```
 * This v2 export will be removed in a future major version.
 */
export function createServerHarborClient(options: ServerHarborClientOptions): HarborClient {
  if (process.env.NODE_ENV === 'production') {
    return createIamHarborClient({
      ...options,
      audience: options.audience ?? options.baseURL,
    });
  }

  return createHarborClient(options);
}
